File: c#/DisgustingEncryption/AES/Aes_128_Cbc_001.cs
namespace DisgustingEncryption
{
    using System;
    using System.Security.Cryptography;
    using System.Text;

    public class Aes_128_Cbc_001
    {
        public static void Show()
        {
            var key1 = "12345678";
            var str1 = "123";

            var enc1 = AesEncrypt(str1, key1);
            // 3gVLeGnili1JBTYLHAk8pQ==
            Console.WriteLine("after encrypt");
            Console.WriteLine(enc1);

            var dec1 = AesDecrypt(enc1, key1);
            Console.WriteLine("after decrypt");
            Console.WriteLine(dec1);


            var key2 = "1234567812345678";
            var str2 = "你好abcd1234";

            var enc2 = AesEncrypt(str2, key2);
            // Qkz+MXCIESJZVgHJffouTQ==
            Console.WriteLine("after encrypt");
            Console.WriteLine(enc2);

            var dec2 = AesDecrypt(enc2, key2);
            Console.WriteLine("after decrypt");
            Console.WriteLine(dec2);
        }


        public static string AesEncrypt(string toEncrypt, string key)
        {
            byte[] keyArray = SHA256(key);
            byte[] toEncryptArray = Encoding.UTF8.GetBytes(toEncrypt);
            byte[] iv = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, };

            RijndaelManaged rDel = new RijndaelManaged
            {
                Key = keyArray,
                IV = iv,
                Mode = CipherMode.CBC,
                Padding = PaddingMode.PKCS7,
                BlockSize = 128
            };

            ICryptoTransform cTransform = rDel.CreateEncryptor();

            byte[] resultArray = cTransform.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length);

            return Convert.ToBase64String(resultArray, 0, resultArray.Length);
        }

        private static byte[] SHA256(string str)
        {
            byte[] SHA256Data = Encoding.UTF8.GetBytes(str);
            SHA256Managed Sha256 = new SHA256Managed();
            byte[] by = Sha256.ComputeHash(SHA256Data);
            return by;
        }

        public static string AesDecrypt(string toDecrypt, string key)
        {
            byte[] keyArray = SHA256(key);
            
            byte[] toDecryptArray = Convert.FromBase64String(toDecrypt); ;
            byte[] iv = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, };

            RijndaelManaged rDel = new RijndaelManaged
            {
                Key = keyArray,
                IV = iv,
                Mode = CipherMode.CBC,
                Padding = PaddingMode.PKCS7,
                BlockSize = 128
            };

            ICryptoTransform cTransform = rDel.CreateDecryptor();

            byte[] resultArray = cTransform.TransformFinalBlock(toDecryptArray, 0, toDecryptArray.Length);

            return Encoding.UTF8.GetString(resultArray);
        }

    }
}

File: python/disgustingEncryption/AES/aes_128_cbc_001.py
from Crypto.Cipher import AES
import base64
import hashlib

class aes_128_cbc_001:


    def get_sha256(self, data):
        sha256 = hashlib.sha256()
        sha256.update(data.encode("utf-8"))    
        res = sha256.digest()        
        return res

    def pkcs7padding(self, text):

        bs = AES.block_size 
        length = len(text)
        bytes_length = len(bytes(text, encoding='utf-8'))

        padding_size = length if(bytes_length == length) else bytes_length
        padding = bs - padding_size % bs
        
        padding_text = chr(padding) * padding
        return text + padding_text


    def pkcs7unpadding(self, text):
        length = len(text)
        unpadding = ord(text[length-1])
        return text[0:length-unpadding]        

    def aes_encrypt(self, content, key):

        key_bytes = self.get_sha256(key)
        iv = "\0".encode("utf-8") * 16

        aes = AES.new(key_bytes, AES.MODE_CBC, iv)

        content_padding = self.pkcs7padding(content)

        encrypt_bytes = aes.encrypt(bytes(content_padding, encoding='utf-8'))

        result = str(base64.b64encode(encrypt_bytes), encoding='utf-8')
        return result

    def aes_decrypt(self, content, key):

        key_bytes = self.get_sha256(key)
        iv = "\0".encode("utf-8") * 16

        aes = AES.new(key_bytes, AES.MODE_CBC, iv)    

        encrypt_bytes = base64.b64decode(content)        
        decrypt_bytes = aes.decrypt(encrypt_bytes)  

        result = str(decrypt_bytes, encoding='utf-8')
        result = self.pkcs7unpadding(result)
        return result 

    def show(self):
        mystr1 = "123"
        mykey1 = "12345678"

        enc1 = self.aes_encrypt(mystr1, mykey1)
        # 3gVLeGnili1JBTYLHAk8pQ==
        print("after encrypt")
        print(enc1)        

        dec1 = self.aes_decrypt(enc1, mykey1)
        print("after decrypt")
        print(dec1)        

        mystr2 = "你好abcd1234"
        mykey2 = "1234567812345678"

        
        enc2 = self.aes_encrypt(mystr2, mykey2)        
        # Qkz+MXCIESJZVgHJffouTQ==
        print("after encrypt")
        print(enc2)        

        dec2 = self.aes_decrypt(enc2, mykey2)
        print("after decrypt")
        print(dec2)        


